Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -0,0 +1,238 @@
+/// Copyright 2018 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#include "SantaPrefixTree.h"
+
+#include "SNTLogging.h"
+
+#include <libkern/c++/OSArray.h>
+#include <libkern/c++/OSNumber.h>
+#include <libkern/c++/OSSet.h>
+
+SantaPrefixTree::SantaPrefixTree() {
+  root_ = new SantaPrefixNode();
+  ++node_count_;
+
+  spt_lock_grp_attr_ = lck_grp_attr_alloc_init();
+  spt_lock_grp_ = lck_grp_alloc_init("santa-prefix-tree-lock", spt_lock_grp_attr_);
+  spt_lock_attr_ = lck_attr_alloc_init();
+  spt_lock_ = lck_rw_alloc_init(spt_lock_grp_, spt_lock_attr_);
+}
+
+IOReturn SantaPrefixTree::AddPrefix(const char *prefix, uint32_t *node_count) {
+  LOGD("Trying to add prefix: %s", prefix);
+
+  // len is the number of bytes (not necessarily the number of characters) representing the string.
+  size_t len = strlen(prefix);
+
+  // Have we have created a new branch in the tree?
+  auto branched = false;
+
+  // Grab a shared lock until a new branch is required.
+  lck_rw_lock_shared(spt_lock_);
+
+  SantaPrefixNode *node = root_;
+  for (int i = 0; i < len; ++i) {
+    // If there is a node in the path that is considered a prefix, stop adding.
+    // For our purposes we only care about the shortest path that matches.
+    if (node->isPrefix) break;
+
+    // Only process a byte at a time.
+    uint8_t value = prefix[i];
+
+    // Create the child if it does not exist.
+    if (!node->children[value]) {
+      // Upgrade the shared lock.
+      // If the upgrade fails, the shared lock is released.
+      if (!lck_rw_lock_shared_to_exclusive(spt_lock_)) {
+        // Grab a new exclusive lock.
+        lck_rw_lock_exclusive(spt_lock_);
+      }
+
+      // Is there enough room for the rest of the prefix?
+      if (!branched && (node_count_ + (len - i)) > kMaxNodes) {
+        LOGE("Prefix tree is full, can not add: %s", prefix);
+        if (node_count) *node_count = node_count_;
+        lck_rw_unlock_exclusive(spt_lock_);
```

Comment:
Unless I'm mistaken[1] there's a potential threading bug here where 2 threads are adding prefixes at the same time and the count goes over kMaxNodes but part of a prefix has been added - it's not a huge problem as the partial prefix won't end with isPrefix=true so it'll never actually trigger anything but the nodes will be taking up space in the tree.

1: I'm not sure but the branched flag _might_ prevent this

Reply:
Looking into it.

Reply:
I think I fixed this by adding the rest of the prefix to the tree when branching.